### PR TITLE
Combine Headers Middleware

### PR DESCRIPTION
**Pull request by Vlix**
Before submitting your PR, check that you've:

- [x] Bumped the version number
- [x] Documented new APIs with [Haddock markup](https://www.haskell.org/haddock/doc/html/index.html)
- [x] Added [`@since` declarations](http://haskell-haddock.readthedocs.io/en/latest/markup.html#since) to the Haddock

After submitting your PR:

- [x] Update the Changelog.md file with a link to your PR
- [x] Check that CI passes (or if it fails, for reasons unrelated to your change, like CI timeouts)

---

Basically a `Middleware` to help in cases like #588 .

### CODE REVIEW

**Comment by Vlix** [comment]
Note to self: don't forget to add tests for the `KeepOnly` handle type.

**Comment by Vlix** [comment]
Note to self: next to the `KeepOnly` test, also add tests for the "slightly bad" format headers (ending in a comma+spaces)

**Comment by MaxGabriel** [comment]
I'm not super active on yesod anymore, probably will not end up reviewing this @Vlix

File: wai-extra/Network/Wai/Middleware/CombineHeaders.hs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+/are/ combined. (e.g. applications might only
+check the first \"Accept\" header and fail, while
+there might be another one that would match)
+    -}
+module Network.Wai.Middleware.CombineHeaders
+    ( combineHeaders
+    , defaultCombineSettings
+    , setHeader
+    , removeHeader
+    , CombineSettings (..)
```
Comment (snoyberg) [inline wai-extra/Network/Wai/Middleware/CombineHeaders.hs:21]: By exposing the data constructors for these two types, it means that any additions in the future will be breaking changes. I'd rather follow the smart constructor/settings type approach.

Comment (Vlix) [reply wai-extra/Network/Wai/Middleware/CombineHeaders.hs:21]: Will change this next time I can focus on this PR 👍

Comment (Vlix) [reply wai-extra/Network/Wai/Middleware/CombineHeaders.hs]: Made new commit that added the smart constructor approach.

File: wai-extra/Network/Wai/Header.hs [hunk 2]
```diff
@@ -2,6 +2,7 @@
 -- | Some helpers for dealing with WAI 'Header's.
 module Network.Wai.Header
     ( contentLength
+    , dropWhileEnd
```
Comment (snoyberg) [inline wai-extra/Network/Wai/Header.hs:5]: This seems more reasonable for a shared, unexposed utils module.

File: wai-extra/Network/Wai/Middleware/CombineHeaders.hs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+/are/ combined. (e.g. applications might only
+check the first \"Accept\" header and fail, while
+there might be another one that would match)
+    -}
+module Network.Wai.Middleware.CombineHeaders
+    ( combineHeaders
+    , CombineSettings
+    , defaultCombineSettings
+    , HeaderMap
+    , HandleType (..)
```
Comment (snoyberg) [inline wai-extra/Network/Wai/Middleware/CombineHeaders.hs:21]: Sorry, I had intended for the smart contract comment to apply to this type too.

Comment (Vlix) [reply wai-extra/Network/Wai/Middleware/CombineHeaders.hs]: Done :)

File: wai-extra/Network/Wai/Header.hs (hunk 2 above)
Comment (Vlix) [reply wai-extra/Network/Wai/Header.hs]: Done :)

**Review by snoyberg — APPROVED** [review]
Thanks!
